pubky-homeserver/src/data_directory/config_toml.rs

```diff
@@ -0,0 +1,267 @@
+pub struct DriveToml {
+    /// The port on which the Pubky TLS Drive API will listen.
+    #[serde(default = "default_pubky_drive_listen_socket")]
+    pub pubky_listen_socket: SocketAddr,
+    /// The port on which the regular http API will listen.
+    #[serde(default = "default_icann_drive_listen_socket")]
+    pub icann_listen_socket: SocketAddr,
+    /// Optional domain name of the regular http API.
+    #[serde(default = "default_icann_drive_domain")]
+    pub icann_domain: Option<String>,
+}
+
+fn default_icann_drive_listen_socket() -> SocketAddr {
+    let ip = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));
+    let port = 6286;
+    SocketAddr::from((ip, port))
+}
+
+fn default_icann_drive_domain() -> Option<String> {
+    None
+}
+
+/// All configuration related to the admin API
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct AdminToml {
+    /// The socket on which the admin API will listen.
+    #[serde(default = "default_admin_listen_socket")]
+    pub listen_socket: SocketAddr,
+    /// The password for the admin API.
+    #[serde(default = "default_admin_password")]
+    pub admin_password: String,
+}
+
+fn default_admin_password() -> String {
+    "admin".to_string()
+}
+
+fn default_admin_listen_socket() -> SocketAddr {
+    let ip = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));
+    let port = 6288;
+    SocketAddr::from((ip, port))
+}
+
+/// All configuration related to the admin API
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct GeneralToml {
+    /// The mode of the signup.
+    #[serde(
+        default = "default_signup_mode",
+        deserialize_with = "validate_signup_mode"
+    )]
+    pub signup_mode: String,
+}
+
+fn default_signup_mode() -> String {
+    "token_required".to_string()
+}
+
+/// The error that can occur when reading the config file
+#[derive(Debug, thiserror::Error)]
+pub enum ConfigReadError {
+    /// The config file not found
+    #[error("Config file not found. {0}")]
+    ConfigFileNotFound(#[from] std::io::Error),
+    /// The config file is not valid    
+    #[error("Config file is not valid. {0}")]
+    ConfigFileNotValid(#[from] toml::de::Error),
+}
+
+/// The main server configuration
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct ConfigToml {
+    /// The configuration for the general settings.
+    pub general: GeneralToml,
+    /// The configuration for the drive files.
+    pub drive: DriveToml,
+    /// The configuration for the admin API.
+    pub admin: AdminToml,
+    /// The configuration for the pkdns.
+    pub pkdns: PkdnsToml,
+}
+
+fn validate_signup_mode<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let s: String = String::deserialize(deserializer)?;
+    match s.as_str() {
+        "open" | "token_required" => Ok(s),
+        _ => Err(serde::de::Error::custom(
+            "signup_mode must be either \"open\" or \"token_required\"",
+        )),
+    }
+}
```

Comment:
Just noticed we could use the existing `SignupMode` Enum's `TryFrom` trait for validation here, so if we were to add more variants in the future, we don't need to remember to come here again to hardcode a new `str` . 
```suggestion
fn validate_signup_mode<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let s: String = String::deserialize(deserializer)?;
    SignupMode::try_from(s.clone()).map_err(serde::de::Error::custom)?;
    Ok(s)
}
```
(requires importing `use crate::core::SignupMode;` )